lte/gateway/c/oai/tasks/amf/amf_app_ue_context_and_proc.h

```diff
@@ -0,0 +1,1001 @@
+   * false, due to "Inactivity of UE including lack of periodic TAU"
+   */
+  bool ppf;
+
+#define SUBSCRIPTION_UNKNOWN false
+#define SUBSCRIPTION_KNOWN true
+  bool subscription_known;
+  ambr_t used_ambr;
+  subscriber_status_t subscriber_status;
+  network_access_mode_t network_access_mode;
+  bool path_switch_req;
+};
+
+/** @class ue_m5gmm_context_s
+ *  @brief Useful parameters to know in AMF application layer. They are set
+ * according to 3GPP TS.23.518 #6.1.6.2.25
+ */
+class ue_mm_context {
+ public:
+  ue_mm_context() {}
+  ~ue_mm_context() {}
+
+  /* msisdn: The basic MSISDN of the UE. The presence is dictated by its storage
+   *         in the UDM, set by N8 UPDATE LOCATION ANSWER
+   */
+  std::string imsi;
+  bool supi_UnauthInd;
+  std::string pei;
+  uint64_t udmGroupId;   // NfGroupId
+  uint64_t ausfGroupId;  // NfGroupId;
+  std::string routingIndicator;
+  std::string drxParameter;
+  std::string subRfsp;
+  uint32_t usedRfsp;  // RfspIndex type;
+  ambr_t subUeAmbr;
+  bool smsSupport;
+  std::string smsfId;  // NfInstanceId type
+  std::string
+      seafData;  // SeafData which will come while AUSF communication for AUTN.
+  std::string pcfId;                   // NfInstanceId
+  std::string pcfAmPolicyUri;          // Uri
+  std::string pcfUePolicyUri;          // Uri
+  std::string hpcfId;                  // NfInstanceId
+  std::string restrictedRatList;       // array(RatType)
+  std::string forbiddenAreaList;       // array(Area)
+  std::string serviceAreaRestriction;  // ServiceAreaRestriction
+  std::string restrictedCnList;        // array(CoreNetworkType)
+  std::string eventSubscriptionList;   // array(AmfEventSubscription)
+  std::string mmContextList;           // array(MmContext)
+  std::string sessionContextList;      // array(PduSessionContext)
+  std::string traceData;               // TraceData
+
+  /* SCTP assoc id */
+  sctp_assoc_id_t sctp_assoc_id_key;
+
+  /* gNB UE NGAP ID,  Unique identity the UE within gNodeB */
+  gnb_ue_ngap_id_t gnb_ue_ngap_id : 24;
+  gnb_ngap_id_key_t gnb_ngap_id_key;
+
+  /* MME UE S1AP ID, Unique identity the UE within MME */
+  amf_ue_ngap_id_t amf_ue_ngap_id;
+};
+
+class amf_app_ue_context : public amf_ue_context_t, public ue_m5gmm_context_s {
+ public:
+  amf_app_ue_context() {}
+  ~amf_app_ue_context() {}
+  // check & create state information.
+  int amf_insert_ue_context(
+      const amf_ue_context_t* amf_ue_context,
+      const ue_m5gmm_context_s* ue_context_p);
+  amf_ue_ngap_id_t amf_app_ctx_get_new_ue_id(
+      amf_ue_ngap_id_t* amf_app_ue_ngap_id_generator_p);
+  // Notify NGAP about the mapping between amf_ue_ngap_id and
+  // sctp assoc id + gnb_ue_ngap_id
+  void notify_ngap_new_ue_amf_ngap_id_association(
+      const ue_m5gmm_context_s* ue_context_p);
+  void amf_remove_ue_context(
+      amf_ue_context_t* const amf_ue_context,
+      ue_m5gmm_context_s* const ue_context_p);
+};
+
+ue_m5gmm_context_s* amf_create_new_ue_context(void);
```

Comment:
provide a comment for each and every function about its purpose.

Reply:
resolved